### ISSUE TITLE
Cannot get FFT method to work for electrostatic solver

### ISSUE THREAD

**Issue by jwestern** [ECP-WarpX/WarpX#5070]
The electrostatic solver has two methods to choose from, "Multigrid" and "FFT." I want to use FFT. I follow the documentation, which says that FFT only works:

1. in 3D
2. with "open" boundary conditions for the field
3. when WarpX is compiled with the flag `-DWarpX_FFT=ON`

I am using GPU execution, and no mesh refinement. Yet my runs quietly try to use the multigrid method instead of FFT, and then crash with an error complaining that the boundary conditions aren't suitable for multigrid:

```
### ERROR   : Field boundary conditions have to be either periodic, PEC or
#            neumann when using the electrostatic Multigrid solver, but they
#            are pml
#            (/global/homes/j/jwestern/src/warpx/Source/FieldSolver/ElectrostaticSolver.cpp:1057)
```

Does anyone know what's going on?

Thanks!

**Comment by archermarx** [comment]
Can you post your input file, or at least the parts pertaining to how you define the ES solver?

**Comment by jwestern** [comment]
Thanks for your reply. Below are the parts pertaining to defining the ES solver and passing it to the simulation. I'm retaining all of the keyword arguments passed, in case something I consider ineffectual is actually relevant:

```
from pywarpx import picmi

grid = picmi.Cartesian3DGrid(
    number_of_cells=[128, 128, 128],
    lower_bound=[-1, -1, -1],
    upper_bound=[1, 1, 1],
    bc_xmin='open',
    bc_xmax='open,
    bc_ymin='open',
    bc_ymax='open',
    bc_zmin='open',
    bc_zmax='open',
    lower_boundary_conditions_particles=['absorbing', 'absorbing', 'absorbing'],
    upper_boundary_conditions_particles=['absorbing', 'absorbing', 'absorbing'],
    refined_regions=[],
    warpx_max_grid_size=None,
    warpx_max_grid_size_x=None,
    warpx_max_grid_size_y=None,
    warpx_max_grid_size_z=None,
    warpx_blocking_factor=None,
    warpx_blocking_factor_x=None,
    warpx_blocking_factor_y=None,
    warpx_blocking_factor_z=None,
)

solver = picmi.ElectrostaticSolver(
    grid=grid,
    method='FFT',
    required_precision=1e-6,
    warpx_self_fields_verbosity = 0,
)

sim = picmi.Simulation(
    solver=solver,
    max_steps=100,
    verbose=0,
    time_step_size=1e-10,
    warpx_collisions=collisions if collisions else None,
    warpx_amrex_use_gpu_aware_mpi=True,
    warpx_sort_particles_for_deposition=True,
    warpx_numprocs=[2, 2, 1],
    warpx_used_inputs_file='warpx_used_inputs_file',
    warpx_checkpoint_signals=['HUP'],
    warpx_amr_restart=None,
    warpx_amr_check_input=True,
)
```

Also, in case there's a problematic option passed, here's my CMAKE like where the FFT option is passed:

`CLICOLOR=0 cmake -DCMAKE_RULE_MESSAGES:BOOL=OFF -S . -B build_pm_"$_mode"_py -DWarpX_COMPUTE=CUDA -DWarpX_FFT=ON -DWarpX_HEFFTE=ON -DWarpX_QED_TABLE_GEN=ON -DWarpX_APP=OFF -DWarpX_PYTHON=ON -DWarpX_EB=OFF -DWarpX_DIMS="1;2;RZ;3"`

**Comment by archermarx** [comment]
OK, I think I found the problem. In the PICMI interface, the `method` argument is not read, so it will always use the MLMG method.

#### `WarpX/Python/pywarpx/picmi.py: 1331`
```python
class ElectrostaticSolver(picmistandard.PICMI_ElectrostaticSolver):
    """
    See `Input Parameters <https://warpx.readthedocs.io/en/latest/usage/parameters.html>`__ for more information.

    Parameters
    ----------
    warpx_relativistic: bool, default=False
        Whether to use the relativistic solver or lab frame solver

    warpx_absolute_tolerance: float, default=0.
        Absolute tolerance on the lab frame solver

    warpx_self_fields_verbosity: integer, default=2
        Level of verbosity for the lab frame solver
    """
    def init(self, kw):
        self.relativistic = kw.pop('warpx_relativistic', False)
        self.absolute_tolerance = kw.pop('warpx_absolute_tolerance', None)
        self.self_fields_verbosity = kw.pop('warpx_self_fields_verbosity', None)
        self.magnetostatic = kw.pop('warpx_magnetostatic', False)

    def solver_initialize_inputs(self):

        self.grid.grid_initialize_inputs()

        if self.relativistic:
            pywarpx.warpx.do_electrostatic = 'relativistic'
        else:
            if self.magnetostatic:
                pywarpx.warpx.do_electrostatic = 'labframe-electromagnetostatic'
            else:
                pywarpx.warpx.do_electrostatic = 'labframe'
            pywarpx.warpx.self_fields_required_precision = self.required_precision
            pywarpx.warpx.self_fields_absolute_tolerance = self.absolute_tolerance
            pywarpx.warpx.self_fields_max_iters = self.maximum_iterations
            pywarpx.warpx.self_fields_verbosity = self.self_fields_verbosity
            pywarpx.boundary.potential_lo_x = self.grid.potential_xmin
            pywarpx.boundary.potential_lo_y = self.grid.potential_ymin
            pywarpx.boundary.potential_lo_z = self.grid.potential_zmin
            pywarpx.boundary.potential_hi_x = self.grid.potential_xmax
            pywarpx.boundary.potential_hi_y = self.grid.potential_ymax
            pywarpx.boundary.potential_hi_z = self.grid.potential_zmax
```

**Comment by archermarx** [comment]
I'll try and put together a patch today.

**Comment by archermarx** [comment]
The problem seems to be that the PICMI `ElectrostaticSolver` class does not initialize the `pywarpx.poisson_solver` variable. I can try and write a patch in a bit

```python
class ElectrostaticSolver(picmistandard.PICMI_ElectrostaticSolver):
    """
    See `Input Parameters <https://warpx.readthedocs.io/en/latest/usage/parameters.html>`__ for more information.

    Parameters
    ----------
    warpx_relativistic: bool, default=False
        Whether to use the relativistic solver or lab frame solver

    warpx_absolute_tolerance: float, default=0.
        Absolute tolerance on the lab frame solver

    warpx_self_fields_verbosity: integer, default=2
        Level of verbosity for the lab frame solver
    """
    def init(self, kw):
        self.relativistic = kw.pop('warpx_relativistic', False)
        self.absolute_tolerance = kw.pop('warpx_absolute_tolerance', None)
        self.self_fields_verbosity = kw.pop('warpx_self_fields_verbosity', None)
        self.magnetostatic = kw.pop('warpx_magnetostatic', False)

    def solver_initialize_inputs(self):

        self.grid.grid_initialize_inputs()

        if self.relativistic:
            pywarpx.warpx.do_electrostatic = 'relativistic'
        else:
            if self.magnetostatic:
                pywarpx.warpx.do_electrostatic = 'labframe-electromagnetostatic'
            else:
                pywarpx.warpx.do_electrostatic = 'labframe'
            pywarpx.warpx.self_fields_required_precision = self.required_precision
            pywarpx.warpx.self_fields_absolute_tolerance = self.absolute_tolerance
            pywarpx.warpx.self_fields_max_iters = self.maximum_iterations
            pywarpx.warpx.self_fields_verbosity = self.self_fields_verbosity
            pywarpx.boundary.potential_lo_x = self.grid.potential_xmin
            pywarpx.boundary.potential_lo_y = self.grid.potential_ymin
            pywarpx.boundary.potential_lo_z = self.grid.potential_zmin
            pywarpx.boundary.potential_hi_x = self.grid.potential_xmax
            pywarpx.boundary.potential_hi_y = self.grid.potential_ymax
            pywarpx.boundary.potential_hi_z = self.grid.potential_zmax
```

**Comment by RemiLehe** [comment]
@jwestern Could you confirm that the latest version of the `development` branch of WarpX fixes the aboved-mentioned issue? (now that #5081 is merged)

**Comment by jwestern** [comment]
Yes, it appears to have worked (my run doesn't crash anymore, and the output says `Poisson solver:       | FFT-based`)

Thank you both!
